src/teamster/kipptaf/performance_management/assets.py

```diff
@@ -0,0 +1,206 @@
+    "so3",
+    "so4",
+    "so5",
+    "so6",
+    "so7",
+    "so8",
+]
+
+
+def get_iqr_outliers(df: pandas.DataFrame):
+    q1 = numpy.percentile(df["overall_score"], 25)
+    q3 = numpy.percentile(df["overall_score"], 75)
+
+    iqr = q3 - q1  # type: ignore
+
+    outliers_array = numpy.where(
+        (df["overall_score"] < q1 - 1.5 * iqr) | (df["overall_score"] > q3 + 1.5 * iqr)
+    )[0]
+
+    df["is_iqr_outlier"] = df["observer_employee_number"].isin(
+        df.loc[outliers_array]["observer_employee_number"].tolist()
+    )
+
+    return df
+
+
+def get_pca(df: pandas.DataFrame):
+    pca = PCA(n_components=2)
+
+    principal_components = pca.fit_transform(X=df[FIT_TRANSFORM_COLUMNS])
+
+    pca_df = pandas.DataFrame(
+        data=principal_components,
+        columns=["pc1", "pc2"],
+    )
+
+    df = pandas.merge(left=df, right=pca_df, left_index=True, right_index=True)
+
+    # percent of variance
+    df["pc1_variance_explained"] = pca.explained_variance_ratio_[0]
+    df["pc2_variance_explained"] = pca.explained_variance_ratio_[1]
+
+    return df
+
+
+def get_dbscan(df):
+    # All time Outlier Detction Using PCA
+    # 0.6 is the historic optimal epsilon, but we may need to adjust
+    outlier_detection = DBSCAN(min_samples=3, eps=0.6)
+
+    clusters = outlier_detection.fit_predict(X=df[["pc1", "pc2"]])
+
+    cluster_df = pandas.DataFrame(data=clusters, columns=["cluster"])
+
+    df = pandas.merge(left=df, right=cluster_df, left_index=True, right_index=True)
+
+    return df
+
+
+def get_isolation_forest(df: pandas.DataFrame):
+    model = IsolationForest(contamination=0.1)  # assuming 10% of the data are outliers
+
+    model.fit(X=df[FIT_TRANSFORM_COLUMNS])
+
+    outliers = model.predict(X=df[FIT_TRANSFORM_COLUMNS])
+
+    tree_df = pandas.DataFrame(data=outliers, columns=["tree_outlier"])
+
+    df = pandas.merge(left=df, right=tree_df, left_index=True, right_index=True)
+
+    return df
+
+
+@asset(
+    key=[CODE_LOCATION, "performance_management", "outlier_detection"],
+    io_manager_key="io_manager_gcs_avro",
+    group_name="performance_management",
+    partitions_def=MultiPartitionsDefinition(
+        {
+            "academic_year": StaticPartitionsDefinition(
+                ["2018", "2019", "2020", "2021", "2022", "2023"]
+            ),
+            "term": StaticPartitionsDefinition(["PM1", "PM2", "PM3"]),
+        }
+    ),
+)
+def outlier_detection(context: AssetExecutionContext, db_bigquery: BigQueryResource):
+    partition_key: MultiPartitionKey = context.partition_key  # type: ignore
+
+    # load data from extract view
+    with db_bigquery.get_client() as bq:
+        bq_client = bq
+
+    dataset_ref = bigquery.DatasetReference(
+        project=bq_client.project, dataset_id="kipptaf_extracts"
+    )
+
+    rows = bq_client.list_rows(
+        table=dataset_ref.table("rpt_python__manager_pm_averages")
+    )
+
+    df_global = rows.to_dataframe()
+
+    df_global.dropna(inplace=True)
```

Comment:
@kevinverhoff I'm noticing that this drops most rows for historic years/terms. Any idea how to get around that?

Reply:
ignore pre-2023 data